File: Python/601-650/Dota2Senate.py
class Solution(object):
    def predictPartyVictory(self, senate):
        """
        :type senate: str
        :rtype: str
        """
        def helper(s, cnt, c):
            r = ''
            for i in s:
                if cnt == 0:
                    c = i
                if i == c:
                    cnt += 1
                    r += i
                else:
                    cnt -= 1
            if r.find('R') == -1:
                return 'Dire'
            if r.find('D') == -1:
                return 'Radiant'
            return helper(r, cnt, c)
        return helper(senate, 0, '')

File: Java/601-650/Dota2Senate.java
class Dota2Senate {
    public String predictPartyVictory(String senate) {
        int count = 1;
        char c = senate.charAt(0);
        StringBuilder sb = new StringBuilder();
        for (int i = 1; i < senate.length(); i++) {
            if (senate.charAt(i) == c) {
                count++;
                sb.append(c);
            } else {
                count--;
            }
            if (count == 0) {
                return predictPartyVictory(senate.substring(i + 1) + c + sb.toString());
            }
        }
        return winner(c);
    }
    String winner(char c) {
        if (c == 'R') return "Radiant";
        return "Dire";
    }
}
